### PR TITLE
Continous query cache moved to open source

### PR DESCRIPTION
**Pull request by vbekiaris**

### CODE REVIEW

File: hazelcast-client/src/test/java/com/hazelcast/client/map/querycache/ClientQueryCacheBasicTest.java [hunk 1]
```diff
@@ -96,7 +96,7 @@ public void setup() {
         if (useNearCache) {
             clientConfig.addNearCacheConfig(new NearCacheConfig()
                     .setName(TEST_MAP_NAME)
-                    .setCacheLocalEntries(true));
+                    .setInvalidateOnChange(true));
```
Comment (jerrinot) [inline hazelcast-client/src/test/java/com/hazelcast/client/map/querycache/ClientQueryCacheBasicTest.java]: do you think this was a bug in a test which was not caught due weak validation in EE?

Comment (vbekiaris) [inline hazelcast-client/src/test/java/com/hazelcast/client/map/querycache/ClientQueryCacheBasicTest.java]: Indeed, `cacheLocalEntries` option does not make sense for a client, EE side validation should also fail. Created an issue for that https://github.com/hazelcast/hazelcast-enterprise/issues/1183

**Comment by vbekiaris** [comment]
Unrelated test failure in `com.hazelcast.client.scheduledexecutor.ClientScheduledExecutorServiceTest.stats_whenMemberOwned`

**Comment by vbekiaris** [comment]
run-lab-run

File: hazelcast/src/main/java/com/hazelcast/map/impl/client/MapPortableHook.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+ */
+public class MapPortableHook implements PortableHook {
+
+    public static final int F_ID = FactoryIdHelper.getFactoryId(MAP_PORTABLE_FACTORY, MAP_PORTABLE_FACTORY_ID);
+
+    public static final int CREATE_ACCUMULATOR_INFO = 1;
+
+    @Override
+    public int getFactoryId() {
+        return F_ID;
```
Comment (vbekiaris) [inline hazelcast/src/main/java/com/hazelcast/map/impl/client/MapPortableHook.java]: While on EE side, this hook was using `FactoryIdHelper.ENTERPRISE_MAP_PORTABLE_FACTORY_ID` as its ID. I have switched it to the (unused until now) `FactoryIdHelper.MAP_PORTABLE_FACTORY_ID`. The `AccumulatorInfo` class is not used in client-member communication, so this change will not break compatibility with older clients.

File: hazelcast/src/main/java/com/hazelcast/map/impl/event/MapEventPublisher.java [hunk 3]
```diff
@@ -57,5 +57,13 @@ void publishEvent(Address caller, String mapName, EntryEventType eventType,
      */
     void hintMapEvent(Address caller, String mapName, EntryEventType eventType, int numberOfEntriesAffected, int partitionId);
 
+    /**
+     * Notifies {@link com.hazelcast.map.QueryCache} subscribers directly, without publishing an event to
+     * other map listeners. This is necessary in certain cases, such as when loading entries into a map.
+     *
+     * @param eventData the event to publish to query caches
+     */
+    void addEventToQueryCache(Object eventData);
```
Comment (vbekiaris) [inline hazelcast/src/main/java/com/hazelcast/map/impl/event/MapEventPublisher.java]: Promoted this method to `MapEventPublisher` interface (was previously declared in enterprise-side `MapEventPublisher` implementation, so a downcast was necessary).

File: hazelcast/src/main/java/com/hazelcast/map/impl/event/MapEventPublisherImpl.java [hunk 4]
```diff
@@ -76,6 +78,9 @@ public MapEventPublisherImpl(MapServiceContext mapServiceContext) {
         } else {
             this.filteringStrategy = new DefaultEntryEventFilteringStrategy(serializationService, mapServiceContext);
         }
+        this.queryCacheEventPublisher = new QueryCacheEventPublisher(filteringStrategy,
```
Comment (vbekiaris) [inline hazelcast/src/main/java/com/hazelcast/map/impl/event/MapEventPublisherImpl.java]: Implementation of methods publishing events to query cache subscribers are gathered in `QueryCacheEventPublisher`.

**Comment by vbekiaris** [comment]
Most classes in this PR have been moved as-is from EE side. Method definitions & implementations have been pulled up from EE subclasses and sometimes integrated into their OSS counterparts in:
- `IMap` and its implementations. `QueryCache` related methods are marked as `@Beta`.
- `MapEventPublisherImpl` (additional hooks for delivering events to continuous query cache subscribers)
- `MapMigrationAwareService` (query cache accumulators are flushed and removed on the source side of a migration)
- `MapPostJoinAwareService` delivers a list of `AccumulatorInfo` for joining members in the `PostJoinMapOperation`
- `MapClientAwareService` takes care of cleaning up `QueryCache` resources on client disconnect.
- Supporting methods in `MapServiceContext` EE side implementation have been pulled up to `MapServiceContextImpl`.

**Comment by vbekiaris** [comment]
run-lab-run

**Comment by vbekiaris** [comment]
Build failed with `The channel to the build job was already closed. This is most probably a temporary test environment failure.` and `Build step 'Publish artifacts to S3 Bucket' marked build as failure. This is most probably a temporary test environment failure.`

**Comment by vbekiaris** [comment]
run-lab-run

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/map/MapMadePublishableMessageTask.java [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+ */
+public class MapMadePublishableMessageTask
+        extends AbstractAllPartitionsMessageTask<EnterpriseMapMadePublishableCodec.RequestParameters> {
+
+    public MapMadePublishableMessageTask(ClientMessage clientMessage, Node node, Connection connection) {
+        super(clientMessage, node, connection);
+    }
+
+    @Override
+    protected EnterpriseMapMadePublishableCodec.RequestParameters decodeClientMessage(ClientMessage clientMessage) {
```
Comment (ahmetmircik) [inline hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/map/MapMadePublishableMessageTask.java]: Can we rename it like this: `EnterpriseMapMadePublishableCodec`-->`MapMadePublishableCodec `?

Comment (vbekiaris) [inline hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/map/MapMadePublishableMessageTask.java]: Yes, according to @asimarslan it is possible, however it requires an update in client protocol templates which I intend to do after EA is released, so we don't mix those changes in this PR.

**Pull request merged**